pyomo/contrib/solver/tests/solvers/test_solvers.py

```diff
@@ -1129,6 +1130,49 @@ def test_mutable_quadratic_objective(
         self.assertAlmostEqual(m.x.value, 0.6962249634573562, 4)
         self.assertAlmostEqual(m.y.value, 0.09227926676152151, 4)
 
+    @parameterized.expand(input=_load_tests(qp_solvers))
+    def test_mutable_quadratic_objective_qp(
+        self, name: str, opt_class: Type[SolverBase], use_presolve: bool
+    ):
+        opt: SolverBase = opt_class()
+        if not opt.available():
+            raise unittest.SkipTest(f'Solver {opt.name} not available.')
+        if any(name.startswith(i) for i in nl_solvers_set):
+            if use_presolve:
+                opt.config.writer_config.linear_presolve = True
+            else:
+                opt.config.writer_config.linear_presolve = False
+        # test issue #3381
+        m = pyo.ConcreteModel()
+
+        m.x1 = pyo.Var(name='x1', domain=pyo.Reals)
+        m.x2 = pyo.Var(name='x2', domain=pyo.Reals)
+
+        m.p = pyo.Param(initialize=1, mutable=True)
+
+        m.obj = pyo.Objective(
+            expr=m.p * m.x1 * m.x1 + m.x2 * m.x2 - m.x1 * m.x2, sense=pyo.minimize
+        )
+
+        m.con1 = pyo.Constraint(expr=m.x1 >= 1)
+        m.con2 = pyo.Constraint(expr=m.x2 >= 1)
+
+        results = opt.solve(m)
+        self.assertAlmostEqual(m.x1.value, 1, places=4)
+        self.assertAlmostEqual(m.x2.value, 1, places=4)
+        self.assertAlmostEqual(results.incumbent_objective, 1, 4)
+
+        m.p.value = 2.0
+        results = opt.solve(m)
+        self.assertAlmostEqual(m.x1.value, 1, places=4)
+        self.assertAlmostEqual(m.x2.value, 1, places=4)
+        self.assertAlmostEqual(results.incumbent_objective, 2, 4)
+
+        del m.x1
+        results = opt.solve(m)
+        self.assertAlmostEqual(m.x2.value, 1, places=4)
+        self.assertAlmostEqual(results.incumbent_objective, 2, 4)
```

Comment:
This does not do what you think. Deleting a variable from a model does is not like removing a column from an LP. The convention in Pyomo is that the model is defined by the set of active components (normally constraints and objectives, but also things like disjunctions). If the objective still uses `x1`, then pyomo deleting `x1` from the model actually has no effect (except that you cannot easily access `x1` anymore). I'll make a minor modification to this test real quick, and then we can get it merged.

Reply:
Thanks for explaining how deleting works= and updating the tests for me !